### DISCUSSION TITLE
Event when alpine updates DOM?

### DISCUSSION THREAD

**Discussion by MichaelMackus** [alpinejs/alpine#2740] [1. Help] [answered]
There are certain libraries that need to be updated when the DOM changes. Is there a way (perhaps an event?) I can hook into this in Alpine.js? 

For example, I'd like to call a function like `lazyLoad.update()` or `tinyMCE.reload()` when alpine adds the new data/items to the DOM.

I suppose one way I could go about this is to create a DOM observer, but I'm wondering if there's a better way (with better performance).

**Comment by MichaelMackus** [comment, marked as answer]
I have tried creating a normal alpine.js component with `Alpine.data` that calls `lazyload()` on `init()`. This works fine normally, and even with new DOM elements. However, when, for example, the data-src changes on an element, lazyLoad() doesn't get called (since the element has already been initialized). The only solution I can think of (other than a DOM observer) is to litter my code with events whenever a lazy load property is changed (not to mention there are other use cases for this other than lazy loading images).

**Reply by SimoTod** [reply]
What changes the data-src attribute? Usually in Alpine the value would be a component property that you bind to the attribute in the scope and you can use either $watch or x-effect to trigger code when it changes. Observing directly a normal dom attribute is not something alpine does out of the box.

**Reply by MichaelMackus** [reply]
To be more specific:

I have a Cart component I load via `x-data="cart"`. This component has a bunch of images, and HTML markup inside that is controlled by the AJAX cart data. This is all working fine, but if I bind the data-src attribute e.g. `:data-src="image"` then whenever `image` changes the data-src gets updated. I now have to update the lazyloader, but I don't want to add code to all my components that do such a thing (there are others this is only one example) since lazyload code is a separate site-wide concern.

I suppose I could add a watcher for `image` in the above example, and then listen to the update event(s) in the lazyloader. I was just wondering if there was a better way, since there could be a lot of properties I'd have to watch. Perhaps this is a good pattern to use whenever an alpine components' data is updated.

**Reply by SimoTod** [reply]
It might be easier with a real example but you don't need to pollute your component, you can use a directive on your image for that kind of stuff. A super simplified version (note, I don't know the syntax of that plugin, it's just a generic example). 
```html
<img :data-src="image" x-init="lazyload($el); $watch('image', () => lazyload($el)" />
```
or
```html
<img :data-src="image" x-effect="image && lazyload($el)" />
```

**Reply by MichaelMackus** [reply]
@SimoTod Thank you... that should actually work very nicely. I suppose building for the CSP target has made me think more component-heavy, there's no reason this can't be just a function in `window`. It is a little tricky to do one-off JS like this with the CSP build, but no reason it shouldn't work that I can see (x-effect is much nicer, hopefully it is supported there).
